### DISCUSSION TITLE
Proposal: Add a BindableProperty incremental source code generator to CommunityToolkit.Maui

### DISCUSSION THREAD

**Discussion by IndianaGary** [CommunityToolkit/Maui#1094] [New Feature Discussions]
This is related to discussion #458 

I have implemented a BindableProperty incremental source code generator along with an analyzer and codefix. It is nearly complete along with unit tests for all three components.

I would like to contribute it to the toolkit and would like to know the next steps in the process.

**Comment by TheCodeTraveler** [comment]
@pictos Perhaps you could help, merging this with the work you've also done creating a BindableProperty source generator

**Reply by IndianaGary** [reply]
@pictos I have uploaded my BindableProperty code generator code and support code to https://github.com/IndianaGary/BindablePropertyProjects. 

In anticipation of offering a contribution and in conformance to the guidelines I attempted to submit a CLA but the link is no good.

**Reply by pictos** [reply]
@IndianaGary I'll take a look ASAP

**Reply by IndianaGary** [reply]
@pictos @brminnick  I've added a couple of more edits for valid arguments and am adding tests. In addition, after a discussion with @333Fred in the Roslyn Discord group, I plan on a refactoring pass to eliminate some duplicated code since I now know how to put both the analyzer and code generator in the same project. The codefix must remain separate as there are conflicting dependencies.

**Reply by pictos** [reply]
@IndianaGary look into [your attribute](https://github.com/IndianaGary/BindablePropertyProjects/blob/main/BindablePropertyAttributes/BindablePropertyAttribute.cs), we would like to use the new C# generic attribute. Here's what I've done on my implementation.

```csharp
[AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
public sealed class TestBindablePropertyAttribute<TReturnType> : Attribute
{
	public string PropertyName { get; } = string.Empty;
	public Type? OwnerType { get; set; }
	public object? DefaultValue { get; set; }
	public string DefaultBindingMode { get; set; } = string.Empty;
	public string ValidadeValueMethodName { get; set; } = string.Empty;
	public string PropertyChangedMethodName { get; set; } = string.Empty;
	public string PropertyChangingMethodName { get; set; } = string.Empty;
	public string CoerceValueMethodName { get; set; } = string.Empty;
	public string DefaultValueCreatorMethodName { get; set; } = string.Empty;

	public TestBindablePropertyAttribute(string propertyName)
	{
		PropertyName = propertyName;
	}
}
```

The usage will be

```csharp
[CommuniBindableProperty.SourceGenerators.TestBindableProperty<string>("AltText", DefaultValue = "batata")]
[CommuniBindableProperty.SourceGenerators.TestBindableProperty<ConsoleApp1.MyColor>("TextColor", DefaultValue = "Colors.Red", OwnerType = typeof(MyColor), DefaultBindingMode = "BindingMode.TwoWay")]
internal partial class Class1 : View
{
	public Class1()
	{

	}
}
```

**Reply by pictos** [reply]
@IndianaGary my version is working already, maybe some small tweaks to adequate with the latest framework changes since has been stale for a long time. There's not too much to do. I just make available all the arguments on `BindableProperty.Create`, the first one is with the required inputs and the rest will use the defaults, the second one is if the user wants to provide all inputs. Yeah, we know it's available on C#11 and that feels more idiomatic, is the same idea for people that uses DI.
We use it at the class level and maybe ctor (not sure) to avoid allocating more memory, one class is fine, but imagine a huge project with a lot of custom controls, we believe that framework shouldn't abuse in terms of memory usage/allocation.

**Reply by IndianaGary** [reply]
Do you have an analyzer and codefix for the generator to catch common errors? That was one of the things I focused on to make usage easier. How about extensive unit tests? I have nearly 30, so far.

Here is a variation on my own approach that incorporates the direction you are going. I show using both with and without generic attributes. Let me know what you think of it.

[BindablePropertyAttribute.txt](https://github.com/CommunityToolkit/Maui/files/10994810/BindablePropertyAttribute.txt)

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
